txn_writer.go

```diff
@@ -230,3 +346,226 @@ fail:
+		l2 = &gps.SimpleLock{}
+	}
+
+	p1, p2 := l1.Projects(), l2.Projects()
+
+	// Check if the slices are sorted already. If they are, we can compare
+	// without copying. Otherwise, we have to copy to avoid altering the
+	// original input.
+	sp1, sp2 := SortedLockedProjects(p1), SortedLockedProjects(p2)
+	if len(p1) > 1 && !sort.IsSorted(sp1) {
+		p1 = make([]gps.LockedProject, len(p1))
+		copy(p1, l1.Projects())
+		sort.Sort(SortedLockedProjects(p1))
+	}
+	if len(p2) > 1 && !sort.IsSorted(sp2) {
+		p2 = make([]gps.LockedProject, len(p2))
+		copy(p2, l2.Projects())
+		sort.Sort(SortedLockedProjects(p2))
+	}
+
+	diff := LockDiff{}
+
+	h1 := l1.InputHash()
+	h2 := l2.InputHash()
+	if !bytes.Equal(h1, h2) {
+		diff.HashDiff = &StringDiff{Previous: string(h1), Current: string(h2)}
+	}
+
+	var i2next int
+	for i1 := 0; i1 < len(p1); i1++ {
+		lp1 := p1[i1]
+		pr1 := lp1.Ident().ProjectRoot
+
+		var matched bool
+		for i2 := i2next; i2 < len(p2); i2++ {
+			lp2 := p2[i2]
+			pr2 := lp2.Ident().ProjectRoot
+
+			switch strings.Compare(string(pr1), string(pr2)) {
+			case 0: // Found a matching project
+				matched = true
+				pdiff := diffProjects(lp1, lp2)
+				if pdiff != nil {
+					diff.Modify = append(diff.Modify, *pdiff)
+				}
+				i2next = i2 + 1 // Don't evaluate to this again
+			case -1: // Found a new project
+				add := buildAddProject(lp2)
+				diff.Add = append(diff.Add, add)
+				i2next = i2 + 1 // Don't evaluate to this again
+				continue        // Keep looking for a matching project
+			case +1: // Project has been removed, handled below
+				break
+			}
+
+			break // Done evaluating this project, move onto the next
+		}
+
+		if !matched {
+			diff.Remove = append(diff.Remove, pr1)
+		}
+	}
+
+	// Anything that still hasn't been evaluated are adds
+	for i2 := i2next; i2 < len(p2); i2++ {
+		lp2 := p2[i2]
+		add := buildAddProject(lp2)
+		diff.Add = append(diff.Add, add)
+	}
+
+	if diff.HashDiff == nil && len(diff.Add) == 0 && len(diff.Remove) == 0 && len(diff.Modify) == 0 {
+		return nil // The locks are the equivalent
+	}
+	return &diff
+}
+
+func buildAddProject(lp gps.LockedProject) LockedProjectDiff {
+	r2, b2, v2 := getVersionInfo(lp.Version())
+	var rev, version, branch *StringDiff
+	if r2 != "" {
+		rev = &StringDiff{Previous: r2, Current: r2}
+	}
+	if b2 != "" {
+		branch = &StringDiff{Previous: b2, Current: b2}
+	}
+	if v2 != "" {
+		version = &StringDiff{Previous: v2, Current: v2}
+	}
+	add := LockedProjectDiff{
+		Name:     lp.Ident().ProjectRoot,
+		Revision: rev,
+		Version:  version,
+		Branch:   branch,
+		Packages: make([]StringDiff, len(lp.Packages())),
+	}
+	for i, pkg := range lp.Packages() {
+		add.Packages[i] = StringDiff{Previous: pkg, Current: pkg}
+	}
+	return add
+}
+
+// diffProjects compares two projects and identifies the differences between them.
+// Returns nil if there are no differences
+// TODO(carolynvs) this should be moved to gps and updated once the gps unexported fields are available to use.
```

Comment:
I've actually been meaning to change gps to export those fields. The thing I was trying to do (make it logically impossible for `LockedProject`s to have nil `Version` data) by not exporting them in the first place was dumb, awkward, and didn't even accomplish the goal.

Doesn't matter too much, though - b/c yes, we'll be promoting this up to gps, anyway. 😄